internal/juju/applications.go

```diff
@@ -466,36 +509,31 @@ func (c applicationsClient) ReadApplication(input *ReadApplicationInput) (*ReadA
 		return nil, fmt.Errorf("failed to get app configuration %v", err)
 	}
 
-	conf := make(map[string]string, 0)
+	conf := make(map[string]ConfigEntry, 0)
 	if returnedConf.ApplicationConfig != nil {
 		for k, v := range returnedConf.ApplicationConfig {
 			// skip the trust value. We have an independent field for that
 			if k == "trust" {
 				continue
 			}
 			// The API returns the configuration entries as interfaces
-			// In the terraform plan we introduce strings...
-			// so we force this conversion
 			aux := v.(map[string]interface{})
 			// set if we find the value key and this is not a default
 			// value.
-			// TODO this returns a dictonary with entries using different
-			// type values. Cast the value to the corresponding type.
-			// indicated in the type field"
-			isDefault, found := aux["source"]
-			if found && isDefault != "default" {
-				if value, found := aux["value"]; found {
-					conf[k] = fmt.Sprintf("%s", value)
+			if value, found := aux["value"]; found {
+				conf[k] = ConfigEntry{
+					Value:     value,
+					IsDefault: aux["source"] == "default",
 				}
 			}
 		}
 		// repeat the same steps for charm config values
 		for k, v := range returnedConf.CharmConfig {
 			aux := v.(map[string]interface{})
-			isDefault, found := aux["source"]
-			if found && isDefault != "default" {
-				if value, found := aux["value"]; found {
-					conf[k] = fmt.Sprintf("%s", value)
+			if value, found := aux["value"]; found {
+				conf[k] = ConfigEntry{
+					Value:     value,
+					IsDefault: aux["source"] == "default",
 				}
```

Comment:
I may be missing something here, but for some config keys won't the CharmConfig will overwrite the `conf[k]` here (that already has the value coming from ApplicationConfig)?

Reply:
Yes. This will happen for example if we import the application from an existing deployment.